src/remotedevDevtools.js

```diff
@@ -0,0 +1,171 @@
+// code taken from
+// https://github.com/zalmoxisus/remotedev/blob/f8256d934316b37a94cd24870fc1d3efcbe7d0b9/src/devTools.js
+// and customized
+import { stringify } from 'jsan';
+import socketCluster from 'socketcluster-client';
+import getHostForRN from 'rn-host-detect';
+
+const defaultSocketOptions = {
+  secure: true,
+  hostname: 'remotedev.io',
+  port: 443,
+  autoReconnect: true,
+  autoReconnectOptions: {
+    randomness: 60000,
+  },
+};
+
+let socket;
+let channel;
+const listeners = {};
+let obtainedUrl = null;
+
+function generateId() {
+  return Math.random().toString(36).substr(2);
+}
+
+function handleMessages(message) {
+  if (!message.payload) message.payload = message.action;
+  Object.keys(listeners).forEach((id) => {
+    if (message.instanceId && id !== message.instanceId) return;
+    if (typeof listeners[id] === 'function') listeners[id](message);
+    else
+      listeners[id].forEach((fn) => {
+        fn(message);
+      });
+  });
+}
+
+function watch() {
+  if (channel) return;
+  socket.emit('login', 'master', (err, channelName) => {
+    if (err) {
+      console.log(err);
+      return;
+    }
+    channel = socket.subscribe(channelName);
+    channel.watch(handleMessages);
+    socket.on(channelName, handleMessages);
+  });
+}
+
+function connectToServer(options) {
+  if (socket) return;
+  let socketOptions;
+  if (options.port) {
+    socketOptions = {
+      port: options.port,
+      hostname: getHostForRN(options.hostname || 'localhost'),
+      secure: !!options.secure,
+    };
+  } else socketOptions = defaultSocketOptions;
+  socket = socketCluster.create(socketOptions);
+  watch();
+}
+
+// obtains ulr from the promise and then performs the callback
+// also, overwrites `optionsObject.hostname`
+// obtained url is stored in `obtainedUrl` variable
+// if it's already obtained, skip fetching
+function obtainUrl(urlPromise, optionsObject, callback) {
```

Comment:
Instead of a function that takes a callback, I think this would make more sense as an async function. If we make `start` an `async` function, we can have it `await urlPromise`, and then `await handleURL(url, options)`. In this system we replace `obtainUrl` with `handleURL`... it's basically the same code, but I renamed it because it's no longer "obtaining" the URL by waiting on `urlPromise` (that would be handled by our new `async` `start` function)